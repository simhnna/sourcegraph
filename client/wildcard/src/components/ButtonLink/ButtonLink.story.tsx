import { Meta, Story } from '@storybook/react'
import React from 'react'

import { BrandedStory } from '@sourcegraph/branded/src/components/BrandedStory'
import webStyles from '@sourcegraph/web/src/SourcegraphWebApp.scss'
import { Grid } from '@sourcegraph/wildcard'

import { ButtonLink } from '.'

const Config: Meta = {
    title: 'wildcard/ButtonLink',

    decorators: [
        story => (
            <BrandedStory styles={webStyles}>{() => <div className="container mt-3">{story()}</div>}</BrandedStory>
        ),
    ],

    parameters: {
        component: ButtonLink,
    },
}

export default Config

export const Simple: Story = () => (
    <>
        <h3>Simple button link</h3>
        <ButtonLink target="_blank" rel="noopener noreferrer" variant="primary" to="http://example.com">
            Click me!
        </ButtonLink>
        <br />
        <h3>Button link with undefined to property</h3>
        <ButtonLink variant="primary" to={undefined}>
            Click me
        </ButtonLink>
        <br />
        <h3>Button link sizes</h3>

        <ButtonLink variant="primary" to="#" size="sm">
            Small Size
        </ButtonLink>
        <br />
        <br />
        <ButtonLink variant="primary" to="#" size="lg">
            Large Size
        </ButtonLink>
        <br />
        <h3>Button link variants</h3>
        <Grid>
            <ButtonLink to="#" variant="primary">
                Primary
            </ButtonLink>
            <ButtonLink to="#" variant="secondary">
                Secondary
            </ButtonLink>
            <ButtonLink to="#" variant="danger">
                Danger
            </ButtonLink>
            <ButtonLink to="#" variant="success">
                Success
            </ButtonLink>
            <ButtonLink to="#" variant="warning">
                Warning
            </ButtonLink>
            <ButtonLink to="#" variant="info">
                Info
            </ButtonLink>
            <ButtonLink to="#" variant="merged">
                Merged
            </ButtonLink>
        </Grid>
    </>
)
