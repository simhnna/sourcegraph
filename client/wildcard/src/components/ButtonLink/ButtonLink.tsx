import * as H from 'history'
import { noop } from 'lodash'
import React, { useCallback, AnchorHTMLAttributes } from 'react'
import { Key } from 'ts-key-enum'

import { RouterLink, AnchorLink } from '@sourcegraph/wildcard'

import { Button } from '..'
import { ButtonProps } from '../Button'

const isSelectKeyPress = (event: React.KeyboardEvent): boolean =>
    event.key === Key.Enter && !event.ctrlKey && !event.shiftKey && !event.metaKey && !event.altKey

export type ButtonLinkProps = Omit<ButtonProps, 'as'> &
    AnchorHTMLAttributes<HTMLAnchorElement> & {
        /** The link destination URL. */
        to?: H.LocationDescriptor

        /**
         * Called when the user clicks or presses enter on this element.
         */
        onSelect?: (event: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>) => void
    }

/**
 * A component that is displayed in the same way, regardless of whether it's a link (with a
 * destination URL) or a button (with a click handler).
 *
 * It is keyboard accessible: unlike `<Link>` or `<a>`, pressing the enter key triggers it.
 */
export const ButtonLink: React.FunctionComponent<ButtonLinkProps> = React.forwardRef(
    ({ className, to, disabled, onSelect = noop, ...rest }, reference) => {
        // We need to set up a keypress listener because <a onclick> doesn't get
        // triggered by enter.
        const onAnchorKeyPress: React.KeyboardEventHandler<HTMLElement> = useCallback(
            event => {
                if (!disabled && isSelectKeyPress(event)) {
                    onSelect(event)
                }
            },
            [onSelect, disabled]
        )

        const onClickPreventDefault: React.MouseEventHandler<HTMLElement> = useCallback(
            event => {
                // Prevent default action of reloading page because of empty href
                event.preventDefault()

                if (disabled) {
                    return false
                }

                return onSelect(event)
            },
            [onSelect, disabled]
        )

        const commonProps = {
            disabled,
            onClick: onSelect,
            onKeyPress: onAnchorKeyPress,
            className,
            ref: reference,
        }

        if (!to || disabled) {
            return (
                <Button
                    {...commonProps}
                    as={AnchorLink}
                    to=""
                    onClick={onClickPreventDefault}
                    onAuxClick={onClickPreventDefault}
                    role="button"
                    {...rest}
                />
            )
        }

        return <Button {...commonProps} as={RouterLink} to={to} {...rest} />
    }
)
