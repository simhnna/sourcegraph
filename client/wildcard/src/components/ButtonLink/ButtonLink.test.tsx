import { render } from '@testing-library/react'
import React from 'react'

import { renderWithRouter } from '@sourcegraph/shared/src/testing/render-with-router'

import { ButtonLink } from './ButtonLink'

describe('<ButtonLink />', () => {
    test('renders correctly btn classes', () => {
        const { asFragment } = render(
            <ButtonLink to="http://example.com" variant="secondary" size="lg">
                Button link
            </ButtonLink>
        )
        expect(asFragment()).toMatchSnapshot()
    })
    test('renders correctly `disabled`', () => {
        const { asFragment } = render(
            <ButtonLink to="http://example.com" variant="secondary" size="lg" disabled={true}>
                Button link
            </ButtonLink>
        )
        expect(asFragment()).toMatchSnapshot()
    })
    test('renders correctly empty `to`', () => {
        const { asFragment } = render(
            <ButtonLink to={undefined} variant="secondary" size="lg">
                Button link
            </ButtonLink>
        )
        expect(asFragment()).toMatchSnapshot()
    })
    test('renders correctly anchor attributes', () => {
        const { asFragment } = renderWithRouter(
            <ButtonLink
                to="https://sourcegraph.com"
                variant="secondary"
                size="lg"
                target="_blank"
                rel="noopener noreferrer"
                data-tooltip="SourceGraph.com"
                data-pressed="true"
            >
                Button link
            </ButtonLink>
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
