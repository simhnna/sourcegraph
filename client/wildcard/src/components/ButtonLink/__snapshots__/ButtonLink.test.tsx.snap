// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonLink /> renders correctly \`disabled\` 1`] = `
<DocumentFragment>
  <a
    class="anchorLink btn btn-secondary btn-lg"
    disabled=""
    href=""
    role="button"
  >
    Button link
  </a>
</DocumentFragment>
`;

exports[`<ButtonLink /> renders correctly anchor attributes 1`] = `
<DocumentFragment>
  <a
    class="anchorLink btn btn-secondary btn-lg"
    data-pressed="true"
    data-tooltip="SourceGraph.com"
    href="https://sourcegraph.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    Button link
  </a>
</DocumentFragment>
`;

exports[`<ButtonLink /> renders correctly btn classes 1`] = `
<DocumentFragment>
  <a
    class="anchorLink btn btn-secondary btn-lg"
    href="http://example.com"
  >
    Button link
  </a>
</DocumentFragment>
`;

exports[`<ButtonLink /> renders correctly empty \`to\` 1`] = `
<DocumentFragment>
  <a
    class="anchorLink btn btn-secondary btn-lg"
    href=""
    role="button"
  >
    Button link
  </a>
</DocumentFragment>
`;
